libraries/AP_HAL_Linux/UDPDevice.cpp

```diff
@@ -17,55 +21,96 @@ UDPDevice::UDPDevice(const char *ip, uint16_t port, bool bcast):
 
 UDPDevice::~UDPDevice()
 {
+    close();
 }
 
 ssize_t UDPDevice::write(const uint8_t *buf, uint16_t n)
 {
-    if (!socket.pollout(0)) {
-        return -1;
-    }
-    if (_connected) {
-        return socket.send(buf, n);
+    if (!_pollable && !open()) {
+        return -EAGAIN;
     }
-    return socket.sendto(buf, n, _ip, _port);
+
+    return _pollable->buffered_write(buf, n);
 }
 
 ssize_t UDPDevice::read(uint8_t *buf, uint16_t n)
 {
-    ssize_t ret = socket.recv(buf, n, 0);
-    if (!_connected && ret > 0) {
-        const char *ip;
-        uint16_t port;
-        socket.last_recv_address(ip, port);
-        _connected = socket.connect(ip, port);
```

Comment:
This doesn't look right. This feature is used so we start the ardupilot broadcasting the heartbeat. The GCS will receive the broadcast, and then send something so the flight stack knows where to stream everything. Ardupilot is always the server, streaming to somewhere.